File: java-webflux/src/main/java/com/ederfmatos/javawebflux/usecase/GetTransaction.java
package com.ederfmatos.javawebflux.usecase;

import com.ederfmatos.javawebflux.entity.Transaction;
import com.ederfmatos.javawebflux.enums.Currency;
import com.ederfmatos.javawebflux.exceptions.TransactionNotFoundException;
import com.ederfmatos.javawebflux.repository.TransactionRepository;
import java.math.BigDecimal;
import lombok.RequiredArgsConstructor;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;

@Component
@RequiredArgsConstructor
public class GetTransaction {
    private final TransactionRepository transactionRepository;

    public Mono<Output> execute(Input input) {
        return transactionRepository.findById(input.id)
                .map(Output::create)
                .switchIfEmpty(Mono.error(TransactionNotFoundException::new));
    }

    public record Input(String id) {
    }

    public record Output(String id, String description, BigDecimal amount, BigDecimal totalAmount, Currency currency) {
        public static Output create(Transaction transaction) {
            return new Output(
                    transaction.id(),
                    transaction.description(),
                    transaction.amount(),
                    transaction.totalAmount(),
                    transaction.currency()
            );
        }
    }
}

File: kotlin-tomcat/src/main/kotlin/com/ederfmatos/kotlintomcat/usecase/GetTransaction.kt
package com.ederfmatos.kotlintomcat.usecase

import com.ederfmatos.kotlintomcat.enums.Currency
import com.ederfmatos.kotlintomcat.exception.TransactionNotFoundException
import com.ederfmatos.kotlintomcat.repository.TransactionRepository
import org.springframework.stereotype.Component
import java.math.BigDecimal

@Component
class GetTransaction(private val transactionRepository: TransactionRepository) {

    fun execute(input: Input): Output {
        val transaction = transactionRepository.findById(input.id) ?: throw TransactionNotFoundException()
        return Output(
            id = transaction.id,
            description = transaction.description,
            amount = transaction.amount,
            currency = transaction.currency,
            totalAmount = transaction.totalAmount
        )
    }

    data class Input(val id: String)
    data class Output(
        val id: String,
        val description: String,
        val amount: BigDecimal,
        val totalAmount: BigDecimal,
        val currency: Currency,
    )
}